// react-bootstrap-typeahead/cjs/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  addCustomOption: true,
  defaultFilterBy: true,
  defaultSelectHint: true,
  getDisplayName: true,
  getHintText: true,
  getInputProps: true,
  getInputText: true,
  getIsOnlyResult: true,
  getMatchBounds: true,
  getMenuItemId: true,
  getOptionLabel: true,
  getOptionProperty: true,
  getStringLabelKey: true,
  getTruncatedOptions: true,
  getUpdatedActiveIndex: true,
  hasOwnProperty: true,
  isSelectable: true,
  isShown: true,
  preventInputBlur: true,
  propsWithBsClassName: true,
  stripDiacritics: true,
  validateSelectedPropChange: true,
  warn: true
};
Object.defineProperty(exports, "addCustomOption", {
  enumerable: true,
  get: function get() {
    return _addCustomOption["default"];
  }
});
Object.defineProperty(exports, "defaultFilterBy", {
  enumerable: true,
  get: function get() {
    return _defaultFilterBy["default"];
  }
});
Object.defineProperty(exports, "defaultSelectHint", {
  enumerable: true,
  get: function get() {
    return _defaultSelectHint["default"];
  }
});
Object.defineProperty(exports, "getDisplayName", {
  enumerable: true,
  get: function get() {
    return _getDisplayName["default"];
  }
});
Object.defineProperty(exports, "getHintText", {
  enumerable: true,
  get: function get() {
    return _getHintText["default"];
  }
});
Object.defineProperty(exports, "getInputProps", {
  enumerable: true,
  get: function get() {
    return _getInputProps["default"];
  }
});
Object.defineProperty(exports, "getInputText", {
  enumerable: true,
  get: function get() {
    return _getInputText["default"];
  }
});
Object.defineProperty(exports, "getIsOnlyResult", {
  enumerable: true,
  get: function get() {
    return _getIsOnlyResult["default"];
  }
});
Object.defineProperty(exports, "getMatchBounds", {
  enumerable: true,
  get: function get() {
    return _getMatchBounds["default"];
  }
});
Object.defineProperty(exports, "getMenuItemId", {
  enumerable: true,
  get: function get() {
    return _getMenuItemId["default"];
  }
});
Object.defineProperty(exports, "getOptionLabel", {
  enumerable: true,
  get: function get() {
    return _getOptionLabel["default"];
  }
});
Object.defineProperty(exports, "getOptionProperty", {
  enumerable: true,
  get: function get() {
    return _getOptionProperty["default"];
  }
});
Object.defineProperty(exports, "getStringLabelKey", {
  enumerable: true,
  get: function get() {
    return _getStringLabelKey["default"];
  }
});
Object.defineProperty(exports, "getTruncatedOptions", {
  enumerable: true,
  get: function get() {
    return _getTruncatedOptions["default"];
  }
});
Object.defineProperty(exports, "getUpdatedActiveIndex", {
  enumerable: true,
  get: function get() {
    return _getUpdatedActiveIndex["default"];
  }
});
Object.defineProperty(exports, "hasOwnProperty", {
  enumerable: true,
  get: function get() {
    return _hasOwnProperty["default"];
  }
});
Object.defineProperty(exports, "isSelectable", {
  enumerable: true,
  get: function get() {
    return _isSelectable["default"];
  }
});
Object.defineProperty(exports, "isShown", {
  enumerable: true,
  get: function get() {
    return _isShown["default"];
  }
});
Object.defineProperty(exports, "preventInputBlur", {
  enumerable: true,
  get: function get() {
    return _preventInputBlur["default"];
  }
});
Object.defineProperty(exports, "propsWithBsClassName", {
  enumerable: true,
  get: function get() {
    return _propsWithBsClassName["default"];
  }
});
Object.defineProperty(exports, "stripDiacritics", {
  enumerable: true,
  get: function get() {
    return _stripDiacritics["default"];
  }
});
Object.defineProperty(exports, "validateSelectedPropChange", {
  enumerable: true,
  get: function get() {
    return _validateSelectedPropChange["default"];
  }
});
Object.defineProperty(exports, "warn", {
  enumerable: true,
  get: function get() {
    return _warn["default"];
  }
});

var _addCustomOption = _interopRequireDefault(require("./addCustomOption"));

var _defaultFilterBy = _interopRequireDefault(require("./defaultFilterBy"));

var _defaultSelectHint = _interopRequireDefault(require("./defaultSelectHint"));

var _getDisplayName = _interopRequireDefault(require("./getDisplayName"));

var _getHintText = _interopRequireDefault(require("./getHintText"));

var _getInputProps = _interopRequireDefault(require("./getInputProps"));

var _getInputText = _interopRequireDefault(require("./getInputText"));

var _getIsOnlyResult = _interopRequireDefault(require("./getIsOnlyResult"));

var _getMatchBounds = _interopRequireWildcard(require("./getMatchBounds"));

Object.keys(_getMatchBounds).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getMatchBounds[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _getMatchBounds[key];
    }
  });
});

var _getMenuItemId = _interopRequireDefault(require("./getMenuItemId"));

var _getOptionLabel = _interopRequireDefault(require("./getOptionLabel"));

var _getOptionProperty = _interopRequireDefault(require("./getOptionProperty"));

var _getStringLabelKey = _interopRequireDefault(require("./getStringLabelKey"));

var _getTruncatedOptions = _interopRequireDefault(require("./getTruncatedOptions"));

var _getUpdatedActiveIndex = _interopRequireDefault(require("./getUpdatedActiveIndex"));

var _hasOwnProperty = _interopRequireDefault(require("./hasOwnProperty"));

var _isSelectable = _interopRequireDefault(require("./isSelectable"));

var _isShown = _interopRequireDefault(require("./isShown"));

var _nodash = require("./nodash");

Object.keys(_nodash).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _nodash[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _nodash[key];
    }
  });
});

var _preventInputBlur = _interopRequireDefault(require("./preventInputBlur"));

var _propsWithBsClassName = _interopRequireDefault(require("./propsWithBsClassName"));

var _size = require("./size");

Object.keys(_size).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _size[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _size[key];
    }
  });
});

var _stripDiacritics = _interopRequireDefault(require("./stripDiacritics"));

var _validateSelectedPropChange = _interopRequireDefault(require("./validateSelectedPropChange"));

var _warn = _interopRequireDefault(require("./warn"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
