// @meteorwallet/sdk/dist/meteor-sdk/src/postMessage/MeteorPostMessenger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMeteorPostMessenger = void 0;
const nanoid_1 = require("nanoid");
const envConfig_1 = require("../envConfig");
const types_dappConnect_1 = require("../../../meteor-common/src/modules_feature/dapp_connect/types_dappConnect");
const query_string_1 = require("query-string");
const error_handling_dappConnect_1 = require("../../../meteor-common/src/modules_feature/dapp_connect/error_handling_dappConnect");
const constants_theme_1 = require("../../../meteor-common/src/modules_app_core/theme/constants_theme");
class ComWindow {
    constructor(connection, listener) {
        const baseWalletUrl = envConfig_1.envConfig.wallet_base_url;
        const url = new URL(baseWalletUrl);
        this.walletOrigin = url.origin;
        if (window.meteorCom == null) {
            this.comType = types_dappConnect_1.EDappActionSource.website_post_message;
            console.log("No extension found. Need to connect to web popup for Meteor communication");
            const queryParams = {
                source: types_dappConnect_1.EDappActionSource.website_post_message,
                connectionUid: connection.uid,
            };
            const w = constants_theme_1.SIGN_POPUP_WIDTH;
            const h = constants_theme_1.SIGN_POPUP_HEIGHT;
            const y = window.top.outerHeight / 2 + window.top.screenY - h / 2;
            const x = window.top.outerWidth / 2 + window.top.screenX - w / 2;
            const newWindow = window.open(`${baseWalletUrl}/connect/${connection.network}/${connection.actionType}?${(0, query_string_1.stringify)(queryParams)}`, "_blank", `popup=1,width=${w},height=${h},top=${y},left=${x}`);
            if (newWindow != null) {
                this.websiteWindow = newWindow;
            }
            else {
                throw new types_dappConnect_1.MeteorActionError({
                    message: "Couldn't open popup window to complete wallet action",
                    endTags: [error_handling_dappConnect_1.EDappActionErrorTag.POPUP_WINDOW_OPEN_FAILED],
                });
            }
            window.addEventListener("message", (event) => listener(event.data));
            this.wasOpened = false;
        }
        else {
            this.comType = types_dappConnect_1.EDappActionSource.extension_injected;
            this.wasOpened = true;
            // console.log("Need to communicate with the extension!");
            window.meteorCom.addMessageDataListener(listener);
        }
    }
    focus() {
        var _a;
        if (this.comType === types_dappConnect_1.EDappActionSource.website_post_message) {
            (_a = this.websiteWindow) === null || _a === void 0 ? void 0 : _a.focus();
        }
    }
    sendMessage(data) {
        var _a, _b;
        if (this.comType === types_dappConnect_1.EDappActionSource.website_post_message) {
            (_a = this.websiteWindow) === null || _a === void 0 ? void 0 : _a.postMessage(data, this.walletOrigin);
        }
        else {
            (_b = window.meteorCom) === null || _b === void 0 ? void 0 : _b.sendMessageData(data);
        }
    }
    isWindowClosed() {
        var _a, _b;
        if (this.comType === types_dappConnect_1.EDappActionSource.website_post_message) {
            return (_b = (_a = this.websiteWindow) === null || _a === void 0 ? void 0 : _a.closed) !== null && _b !== void 0 ? _b : true;
        }
        else {
            return false;
        }
    }
    hasActiveWindow() {
        if (this.comType === types_dappConnect_1.EDappActionSource.website_post_message) {
            return this.websiteWindow != null;
        }
        else {
            return true;
        }
    }
    close() {
        if (this.comType === types_dappConnect_1.EDappActionSource.website_post_message) {
            delete this.websiteWindow;
        }
        else {
        }
    }
}
const pingInterval = 450;
class MeteorPostMessenger {
    constructor() {
        this.connections = [];
        // const baseUrl = "https://dev.wallet.meteorwallet.app";
        const baseUrl = envConfig_1.envConfig.wallet_base_url;
        const url = new URL(baseUrl);
        this.baseWalletUrl = baseUrl;
        this.walletOrigin = url.origin;
        this.listener = (data) => {
            if (data != null) {
                // const data: TClientPostMessageResponse = event.data;
                // console.log("Meteor Post Messenger received event with data: ", data);
                const currentConnection = this.connections.find((con) => con.uid === data.uid);
                if (currentConnection != null) {
                    this.updateConnection(currentConnection.uid, {
                        lastConnection: Date.now(),
                    });
                    if (data.status === types_dappConnect_1.EDappActionConnectionStatus.attempting_reconnect) {
                        this.updateConnection(currentConnection.uid, {
                            status: types_dappConnect_1.EDappActionConnectionStatus.initializing,
                        });
                        this.sendComs();
                    }
                    if (data.status === types_dappConnect_1.EDappActionConnectionStatus.connected &&
                        currentConnection.status ===
                            types_dappConnect_1.EDappActionConnectionStatus.initializing) {
                        this.updateConnection(currentConnection.uid, {
                            status: types_dappConnect_1.EDappActionConnectionStatus.connected,
                        });
                    }
                    if (data.status === types_dappConnect_1.EDappActionConnectionStatus.closed_success) {
                        currentConnection.resolve({
                            success: true,
                            endTags: [],
                            payload: data.payload,
                        });
                        this.updateConnection(currentConnection.uid, {
                            status: types_dappConnect_1.EDappActionConnectionStatus.closed_success,
                        });
                        this.sendComs();
                    }
                    if (data.status === types_dappConnect_1.EDappActionConnectionStatus.closed_fail) {
                        this.updateConnection(currentConnection.uid, {
                            status: types_dappConnect_1.EDappActionConnectionStatus.closed_fail,
                        });
                        this.sendComs();
                        // const error = new MeteorActionError({ endTags: data.endTags });
                        // console.log("Ending failure with error", { ...error });
                        currentConnection.reject(new types_dappConnect_1.MeteorActionError({ endTags: data.endTags }));
                    }
                    if (data.status === types_dappConnect_1.EDappActionConnectionStatus.closed_window) {
                        this.updateConnection(currentConnection.uid, {
                            status: types_dappConnect_1.EDappActionConnectionStatus.closed_window,
                        });
                        currentConnection.reject(new types_dappConnect_1.MeteorActionError({
                            endTags: data.endTags,
                            message: "User closed the window",
                        }));
                    }
                    if ([
                        types_dappConnect_1.EDappActionConnectionStatus.disconnected,
                        types_dappConnect_1.EDappActionConnectionStatus.closed_fail,
                        types_dappConnect_1.EDappActionConnectionStatus.closed_window,
                        types_dappConnect_1.EDappActionConnectionStatus.closed_success,
                    ].includes(data.status)) {
                        this.removeConnection(currentConnection.uid);
                    }
                } /* else {
                  console.warn(`Connection data received but no ID found "${data.uid}"`, data);
                }*/
            }
        };
        // window.addEventListener("message", this.listener);
    }
    removeConnection(uid) {
        this.connections = this.connections.filter((con) => con.uid !== uid);
        if (this.connections.length === 0) {
            delete this.comWindow;
            clearInterval(this.comInterval);
            delete this.comInterval;
        }
    }
    updateConnection(uid, newConnectionProperties) {
        this.connections = this.connections.map((con) => {
            if (con.uid === uid) {
                return {
                    ...con,
                    ...newConnectionProperties,
                    uid: con.uid,
                };
            }
            return con;
        });
    }
    sendComs() {
        var _a;
        if (((_a = this.comWindow) === null || _a === void 0 ? void 0 : _a.hasActiveWindow()) && this.connections.length > 0) {
            if (this.comInterval == null) {
                this.comInterval = setInterval(() => this.sendComs(), pingInterval);
            }
            if (this.comWindow.isWindowClosed()) {
                if (this.comWindow.wasOpened) {
                    for (const con of this.connections) {
                        con.reject(new types_dappConnect_1.MeteorActionError({
                            message: "User closed the window before completing the action",
                            endTags: [
                                error_handling_dappConnect_1.EDappActionErrorTag.INCOMPLETE_ACTION,
                                error_handling_dappConnect_1.EDappActionErrorTag.WINDOW_CLOSED,
                            ],
                        }));
                    }
                    this.connections = [];
                    this.comWindow.close();
                }
                else {
                    console.log("Window is closed, need to allow popup");
                }
            }
            else {
                this.comWindow.wasOpened = true;
                for (const { network, actionType, endTags, status, inputs, uid } of this
                    .connections) {
                    const postMessage = {
                        endTags,
                        actionType: actionType,
                        status,
                        uid,
                        network,
                    };
                    if (status === types_dappConnect_1.EDappActionConnectionStatus.initializing) {
                        postMessage.inputs = inputs;
                    }
                    this.comWindow.sendMessage(postMessage);
                }
            }
        }
        else {
            clearInterval(this.comInterval);
            delete this.comInterval;
        }
    }
    addAndStartConnection(connection) {
        if (this.connections.length > 0) {
            for (const con of this.connections) {
                con.status = types_dappConnect_1.EDappActionConnectionStatus.disconnected;
                con.endTags = [error_handling_dappConnect_1.EDappActionErrorTag.NEW_ACTION_STARTED];
            }
        }
        this.connections.push(connection);
        if (this.comWindow == null || !this.comWindow.hasActiveWindow()) {
            this.comWindow = new ComWindow(connection, this.listener);
        }
        else {
            this.comWindow.focus();
        }
        this.sendComs();
    }
    async connectAndWaitForResponse({ actionType, network, inputs, }) {
        let newConnection = {
            uid: (0, nanoid_1.nanoid)(),
            actionType,
            lastAttemptedConnection: 0,
            lastConnection: 0,
            status: types_dappConnect_1.EDappActionConnectionStatus.initializing,
            promise: undefined,
            reject: undefined,
            resolve: undefined,
            currentPayload: {},
            inputs,
            network,
            endTags: [],
        };
        const promise = new Promise((resolve, reject) => {
            newConnection.resolve = resolve;
            newConnection.reject = (error) => {
                this.connections = this.connections.filter((con) => con.uid !== newConnection.uid);
                reject(error);
            };
        });
        newConnection.promise = promise;
        this.addAndStartConnection(newConnection);
        return await promise;
    }
}
let postMessenger;
function getMeteorPostMessenger() {
    if (postMessenger == null) {
        postMessenger = new MeteorPostMessenger();
    }
    return postMessenger;
}
exports.getMeteorPostMessenger = getMeteorPostMessenger;
