// @tldraw/utils/dist-cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  FileHelpers: () => import_file.FileHelpers,
  MediaHelpers: () => import_media.MediaHelpers,
  PngHelpers: () => import_png.PngHelpers,
  Result: () => import_control.Result,
  annotateError: () => import_error.annotateError,
  assert: () => import_control.assert,
  assertExists: () => import_control.assertExists,
  compact: () => import_array.compact,
  debounce: () => import_debounce.debounce,
  dedupe: () => import_array.dedupe,
  deepCopy: () => import_object.deepCopy,
  exhaustiveSwitchError: () => import_control.exhaustiveSwitchError,
  filterEntries: () => import_object.filterEntries,
  getErrorAnnotations: () => import_error.getErrorAnnotations,
  getFirstFromIterable: () => import_iterable.getFirstFromIterable,
  getHashForObject: () => import_hash.getHashForObject,
  getHashForString: () => import_hash.getHashForString,
  getOwnProperty: () => import_object.getOwnProperty,
  hasOwnProperty: () => import_object.hasOwnProperty,
  isDefined: () => import_value.isDefined,
  isNonNull: () => import_value.isNonNull,
  isNonNullish: () => import_value.isNonNullish,
  isValidUrl: () => import_url.isValidUrl,
  last: () => import_array.last,
  lerp: () => import_number.lerp,
  lns: () => import_hash.lns,
  mapObjectMapValues: () => import_object.mapObjectMapValues,
  minBy: () => import_array.minBy,
  modulate: () => import_number.modulate,
  noop: () => import_function.noop,
  objectMapEntries: () => import_object.objectMapEntries,
  objectMapFromEntries: () => import_object.objectMapFromEntries,
  objectMapKeys: () => import_object.objectMapKeys,
  objectMapValues: () => import_object.objectMapValues,
  omitFromStackTrace: () => import_function.omitFromStackTrace,
  partition: () => import_array.partition,
  promiseWithResolve: () => import_control.promiseWithResolve,
  rafThrottle: () => import_raf.rafThrottle,
  rng: () => import_number.rng,
  rotateArray: () => import_array.rotateArray,
  sortById: () => import_sort.sortById,
  structuredClone: () => import_value.structuredClone,
  throttle: () => import_function.throttle,
  throttledRaf: () => import_raf.throttledRaf
});
module.exports = __toCommonJS(src_exports);
var import_array = require("./lib/array");
var import_control = require("./lib/control");
var import_debounce = require("./lib/debounce");
var import_error = require("./lib/error");
var import_file = require("./lib/file");
var import_function = require("./lib/function");
var import_hash = require("./lib/hash");
var import_iterable = require("./lib/iterable");
var import_media = require("./lib/media");
var import_number = require("./lib/number");
var import_object = require("./lib/object");
var import_png = require("./lib/png");
var import_raf = require("./lib/raf");
var import_sort = require("./lib/sort");
var import_url = require("./lib/url");
var import_value = require("./lib/value");
