// @tldraw/tldraw/dist-cjs/lib/ui/components/RedoButton.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RedoButton_exports = {};
__export(RedoButton_exports, {
  RedoButton: () => RedoButton
});
module.exports = __toCommonJS(RedoButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useActions = require("../hooks/useActions");
var import_useCanRedo = require("../hooks/useCanRedo");
var import_useTranslation = require("../hooks/useTranslation/useTranslation");
var import_Button = require("./primitives/Button");
var import_shared = require("./primitives/shared");
const RedoButton = (0, import_react.memo)(function RedoButton2() {
  const msg = (0, import_useTranslation.useTranslation)();
  const canRedo = (0, import_useCanRedo.useCanRedo)();
  const actions = (0, import_useActions.useActions)();
  const redo = actions["redo"];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      "data-testid": "main.redo",
      icon: redo.icon,
      type: "icon",
      title: `${msg(redo.label)} ${(0, import_shared.kbdStr)(redo.kbd)}`,
      disabled: !canRedo,
      onClick: () => redo.onSelect("quick-actions"),
      smallIcon: true
    }
  );
});
