// @tldraw/tldraw/dist-cjs/lib/ui/components/UndoButton.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var UndoButton_exports = {};
__export(UndoButton_exports, {
  UndoButton: () => UndoButton
});
module.exports = __toCommonJS(UndoButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useActions = require("../hooks/useActions");
var import_useCanUndo = require("../hooks/useCanUndo");
var import_useTranslation = require("../hooks/useTranslation/useTranslation");
var import_Button = require("./primitives/Button");
var import_shared = require("./primitives/shared");
const UndoButton = (0, import_react.memo)(function UndoButton2() {
  const msg = (0, import_useTranslation.useTranslation)();
  const canUndo = (0, import_useCanUndo.useCanUndo)();
  const actions = (0, import_useActions.useActions)();
  const undo = actions["undo"];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      "data-testid": "main.undo",
      icon: undo.icon,
      type: "icon",
      title: `${msg(undo.label)} ${(0, import_shared.kbdStr)(undo.kbd)}`,
      disabled: !canUndo,
      onClick: () => undo.onSelect("quick-actions"),
      smallIcon: true
    }
  );
});
