// @tldraw/tldraw/dist-cjs/lib/utils/assets/assets.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var assets_exports = {};
__export(assets_exports, {
  DEFAULT_ACCEPTED_IMG_TYPE: () => DEFAULT_ACCEPTED_IMG_TYPE,
  DEFAULT_ACCEPTED_VID_TYPE: () => DEFAULT_ACCEPTED_VID_TYPE,
  containBoxSize: () => containBoxSize,
  getResizedImageDataUrl: () => getResizedImageDataUrl,
  isGifAnimated: () => isGifAnimated
});
module.exports = __toCommonJS(assets_exports);
var import_downscale = __toESM(require("downscale"));
var import_getBrowserCanvasMaxSize = require("../../shapes/shared/getBrowserCanvasMaxSize");
var import_is_gif_animated = require("./is-gif-animated");
function containBoxSize(originalSize, containBoxSize2) {
  const overByXScale = originalSize.w / containBoxSize2.w;
  const overByYScale = originalSize.h / containBoxSize2.h;
  if (overByXScale <= 1 && overByYScale <= 1) {
    return originalSize;
  } else if (overByXScale > overByYScale) {
    return {
      w: originalSize.w / overByXScale,
      h: originalSize.h / overByXScale
    };
  } else {
    return {
      w: originalSize.w / overByYScale,
      h: originalSize.h / overByYScale
    };
  }
}
async function getResizedImageDataUrl(dataURLForImage, width, height, opts = {}) {
  let desiredWidth = width * 2;
  let desiredHeight = height * 2;
  const { type = "image/jpeg", quality = 0.92 } = opts;
  const canvasSizes = await (0, import_getBrowserCanvasMaxSize.getBrowserCanvasMaxSize)();
  const aspectRatio = width / height;
  if (desiredWidth > canvasSizes.maxWidth) {
    desiredWidth = canvasSizes.maxWidth;
    desiredHeight = desiredWidth / aspectRatio;
  }
  if (desiredHeight > canvasSizes.maxHeight) {
    desiredHeight = canvasSizes.maxHeight;
    desiredWidth = desiredHeight * aspectRatio;
  }
  if (desiredWidth * desiredHeight > canvasSizes.maxArea) {
    const ratio = Math.sqrt(canvasSizes.maxArea / (desiredWidth * desiredHeight));
    desiredWidth *= ratio;
    desiredHeight *= ratio;
  }
  return await (0, import_downscale.default)(dataURLForImage, desiredWidth, desiredHeight, { imageType: type, quality });
}
const DEFAULT_ACCEPTED_IMG_TYPE = ["image/jpeg", "image/png", "image/gif", "image/svg+xml"];
const DEFAULT_ACCEPTED_VID_TYPE = ["video/mp4", "video/quicktime"];
async function isGifAnimated(file) {
  return await new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onerror = () => reject(reader.error);
    reader.onload = () => {
      resolve(reader.result ? (0, import_is_gif_animated.isAnimated)(reader.result) : false);
    };
    reader.readAsArrayBuffer(file);
  });
}
