// @tldraw/editor/dist-cjs/lib/editor/managers/UserPreferencesManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result)
    __defProp(target, key, result);
  return result;
};
var UserPreferencesManager_exports = {};
__export(UserPreferencesManager_exports, {
  UserPreferencesManager: () => UserPreferencesManager
});
module.exports = __toCommonJS(UserPreferencesManager_exports);
var import_state = require("@tldraw/state");
var import_TLUserPreferences = require("../../config/TLUserPreferences");
class UserPreferencesManager {
  constructor(user, inferDarkMode) {
    this.user = user;
    this.inferDarkMode = inferDarkMode;
  }
  updateUserPreferences = (userPreferences) => {
    this.user.setUserPreferences({
      ...this.user.userPreferences.get(),
      ...userPreferences
    });
  };
  getUserPreferences() {
    return {
      id: this.getId(),
      name: this.getName(),
      locale: this.getLocale(),
      color: this.getColor(),
      isDarkMode: this.getIsDarkMode(),
      animationSpeed: this.getAnimationSpeed(),
      isSnapMode: this.getIsSnapMode()
    };
  }
  /**
   * @deprecated use `getUserPreferences` instead
   */
  get userPreferences() {
    return this.getUserPreferences();
  }
  getIsDarkMode() {
    return this.user.userPreferences.get().isDarkMode ?? (this.inferDarkMode ? (0, import_TLUserPreferences.userPrefersDarkUI)() : false);
  }
  /**
   * @deprecated use `getIsDarkMode` instead
   */
  get isDarkMode() {
    return this.getIsDarkMode();
  }
  getAnimationSpeed() {
    return this.user.userPreferences.get().animationSpeed ?? import_TLUserPreferences.defaultUserPreferences.animationSpeed;
  }
  /**
   * @deprecated use `getAnimationSpeed` instead
   */
  get animationSpeed() {
    return this.getAnimationSpeed();
  }
  getId() {
    return this.user.userPreferences.get().id;
  }
  /**
   * @deprecated use `getId` instead
   */
  get id() {
    return this.getId();
  }
  getName() {
    return this.user.userPreferences.get().name ?? import_TLUserPreferences.defaultUserPreferences.name;
  }
  /**
   * @deprecated use `getName` instead
   */
  get name() {
    return this.getName();
  }
  getLocale() {
    return this.user.userPreferences.get().locale ?? import_TLUserPreferences.defaultUserPreferences.locale;
  }
  /**
   * @deprecated use `getLocale` instead
   */
  get locale() {
    return this.getLocale();
  }
  getColor() {
    return this.user.userPreferences.get().color ?? import_TLUserPreferences.defaultUserPreferences.color;
  }
  /**
   * @deprecated use `getColor` instead
   */
  get color() {
    return this.getColor();
  }
  getIsSnapMode() {
    return this.user.userPreferences.get().isSnapMode ?? import_TLUserPreferences.defaultUserPreferences.isSnapMode;
  }
  /**
   * @deprecated use `getIsSnapMode` instead
   */
  get isSnapMode() {
    return this.getIsSnapMode();
  }
}
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getUserPreferences", 1);
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getIsDarkMode", 1);
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getAnimationSpeed", 1);
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getId", 1);
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getName", 1);
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getLocale", 1);
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getColor", 1);
__decorateClass([
  import_state.computed
], UserPreferencesManager.prototype, "getIsSnapMode", 1);
