// @meteorwallet/sdk/dist/meteor-common/src/modules_feature/dapp_connect/error_handling_dappConnect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExternalActionErrorMessageForEndTag = exports.EDappActionErrorTag = void 0;
var EDappActionErrorTag;
(function (EDappActionErrorTag) {
    EDappActionErrorTag["NEW_ACTION_STARTED"] = "NEW_ACTION_STARTED";
    EDappActionErrorTag["INCOMPLETE_ACTION"] = "INCOMPLETE_ACTION";
    EDappActionErrorTag["NO_ACCOUNTS"] = "NO_ACCOUNTS";
    EDappActionErrorTag["WINDOW_CLOSED"] = "WINDOW_CLOSED";
    EDappActionErrorTag["USER_CANCELLED"] = "USER_CANCELLED";
    EDappActionErrorTag["POPUP_WINDOW_REFUSED"] = "POPUP_WINDOW_REFUSED";
    EDappActionErrorTag["POPUP_WINDOW_OPEN_FAILED"] = "POPUP_WINDOW_OPEN_FAILED";
})(EDappActionErrorTag = exports.EDappActionErrorTag || (exports.EDappActionErrorTag = {}));
const en = {
    [EDappActionErrorTag.WINDOW_CLOSED]: "User closed the window",
    [EDappActionErrorTag.INCOMPLETE_ACTION]: "User didn't complete the action",
    [EDappActionErrorTag.NEW_ACTION_STARTED]: "A new action was started by the user, old action failed",
    [EDappActionErrorTag.POPUP_WINDOW_OPEN_FAILED]: "Popup window failed to open",
    [EDappActionErrorTag.POPUP_WINDOW_REFUSED]: "User refused to allow the popup window to open",
    [EDappActionErrorTag.USER_CANCELLED]: "User cancelled the action",
    [EDappActionErrorTag.NO_ACCOUNTS]: "No Meteor account found- try again after creating one",
};
function getExternalActionErrorMessageForEndTag(tag) {
    var _a;
    return (_a = en[tag]) !== null && _a !== void 0 ? _a : "An unknown error occurred";
}
exports.getExternalActionErrorMessageForEndTag = getExternalActionErrorMessageForEndTag;
