// @meteorwallet/sdk/dist/meteor-common/src/modules_feature/dapp_connect/types_dappConnect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EDappActionConnectionStatus = exports.EDappActionSource = exports.EMeteorInjectedFeature = exports.EMeteorExtensionDirectActionType = exports.EWalletExternalActionStatus = exports.EWalletExternalAction_SignIn_AccessType = exports.EMeteorWalletSignInType = exports.EExternalActionType = exports.MeteorActionError = void 0;
const error_handling_dappConnect_1 = require("./error_handling_dappConnect");
class MeteorActionError extends Error {
    constructor(reason, previousError) {
        var _a, _b;
        super();
        if (reason.message == null) {
            reason.message = (0, error_handling_dappConnect_1.getExternalActionErrorMessageForEndTag)((_b = (_a = reason.endTags) === null || _a === void 0 ? void 0 : _a[reason.endTags.length - 1]) !== null && _b !== void 0 ? _b : "unknown");
        }
        this._reason = reason;
        this.message = reason.message;
        this.name = "MeteorActionError";
        this.cause = previousError;
    }
}
exports.MeteorActionError = MeteorActionError;
var EExternalActionType;
(function (EExternalActionType) {
    EExternalActionType["login"] = "login";
    EExternalActionType["sign"] = "sign";
    EExternalActionType["logout"] = "logout";
    EExternalActionType["verify_owner"] = "verify_owner";
    EExternalActionType["keypom_claim"] = "keypom_claim";
    EExternalActionType["sign_message"] = "sign_message";
})(EExternalActionType = exports.EExternalActionType || (exports.EExternalActionType = {}));
var EMeteorWalletSignInType;
(function (EMeteorWalletSignInType) {
    EMeteorWalletSignInType["ALL_METHODS"] = "ALL_METHODS";
    EMeteorWalletSignInType["SELECTED_METHODS"] = "SELECTED_METHODS";
    EMeteorWalletSignInType["FULL_ACCESS"] = "FULL_ACCESS";
})(EMeteorWalletSignInType = exports.EMeteorWalletSignInType || (exports.EMeteorWalletSignInType = {}));
var EWalletExternalAction_SignIn_AccessType;
(function (EWalletExternalAction_SignIn_AccessType) {
    EWalletExternalAction_SignIn_AccessType["FULL_ACCESS"] = "fullAccess";
    EWalletExternalAction_SignIn_AccessType["LIMITED_ACCESS"] = "limitedAccess";
})(EWalletExternalAction_SignIn_AccessType = exports.EWalletExternalAction_SignIn_AccessType || (exports.EWalletExternalAction_SignIn_AccessType = {}));
var EWalletExternalActionStatus;
(function (EWalletExternalActionStatus) {
    EWalletExternalActionStatus["UNCONFIRMED"] = "UNCONFIRMED";
    EWalletExternalActionStatus["PENDING"] = "PENDING";
    EWalletExternalActionStatus["SUCCESS"] = "SUCCESS";
    EWalletExternalActionStatus["FAILURE"] = "FAILURE";
})(EWalletExternalActionStatus = exports.EWalletExternalActionStatus || (exports.EWalletExternalActionStatus = {}));
var EMeteorExtensionDirectActionType;
(function (EMeteorExtensionDirectActionType) {
    EMeteorExtensionDirectActionType["check_sync_status"] = "check_sync_status";
    EMeteorExtensionDirectActionType["sync_accounts"] = "sync_accounts";
    EMeteorExtensionDirectActionType["open_page"] = "open_page";
})(EMeteorExtensionDirectActionType = exports.EMeteorExtensionDirectActionType || (exports.EMeteorExtensionDirectActionType = {}));
var EMeteorInjectedFeature;
(function (EMeteorInjectedFeature) {
    EMeteorInjectedFeature["open_page"] = "open_page";
    EMeteorInjectedFeature["batch_import"] = "batch_import";
    EMeteorInjectedFeature["sync_check"] = "sync_check";
    EMeteorInjectedFeature["account_sync"] = "account_sync";
})(EMeteorInjectedFeature = exports.EMeteorInjectedFeature || (exports.EMeteorInjectedFeature = {}));
var EDappActionSource;
(function (EDappActionSource) {
    EDappActionSource["website_callback"] = "wcb";
    EDappActionSource["website_post_message"] = "wpm";
    EDappActionSource["website_visit"] = "wv";
    EDappActionSource["extension_injected"] = "ext";
})(EDappActionSource = exports.EDappActionSource || (exports.EDappActionSource = {}));
var EDappActionConnectionStatus;
(function (EDappActionConnectionStatus) {
    EDappActionConnectionStatus["initializing"] = "initializing";
    EDappActionConnectionStatus["connected"] = "connected";
    EDappActionConnectionStatus["attempting_reconnect"] = "attempting_reconnect";
    EDappActionConnectionStatus["disconnected"] = "disconnected";
    EDappActionConnectionStatus["closed_success"] = "closed_success";
    EDappActionConnectionStatus["closed_fail"] = "closed_fail";
    EDappActionConnectionStatus["closed_window"] = "closed_window";
})(EDappActionConnectionStatus = exports.EDappActionConnectionStatus || (exports.EDappActionConnectionStatus = {}));
