// @meteorwallet/sdk/dist/meteor-common/src/modules_feature/accounts/account_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EWalletUserVersion = exports.EProfileStatus = exports.EAccountImportType = exports.EAccountIdentifierType = exports.EGetKeysPermissionType = exports.EAccountFunctionality = exports.EAccountFundingStatus = exports.EAccountProfileType = exports.EDecryptionKeyType = exports.EAccountKeyType = void 0;
var EAccountKeyType;
(function (EAccountKeyType) {
    EAccountKeyType["PW_ENCRYPTED_PRIVATE"] = "PW_ENCRYPTED_PRIVATE";
    EAccountKeyType["HARDWARE"] = "HARDWARE";
    EAccountKeyType["WEB3AUTH"] = "WEB3AUTH";
})(EAccountKeyType = exports.EAccountKeyType || (exports.EAccountKeyType = {}));
var EDecryptionKeyType;
(function (EDecryptionKeyType) {
    EDecryptionKeyType["DIRECT"] = "DIRECT";
    EDecryptionKeyType["HASHED_SHA256_SALTED"] = "HASHED_SHA256_SALTED";
})(EDecryptionKeyType = exports.EDecryptionKeyType || (exports.EDecryptionKeyType = {}));
var EAccountProfileType;
(function (EAccountProfileType) {
    EAccountProfileType["LOCAL"] = "LOCAL";
    EAccountProfileType["NEAR_SOCIAL"] = "NEAR_SOCIAL";
})(EAccountProfileType = exports.EAccountProfileType || (exports.EAccountProfileType = {}));
var EAccountFundingStatus;
(function (EAccountFundingStatus) {
    EAccountFundingStatus["UNFUNDED"] = "UNFUNDED";
    EAccountFundingStatus["FUNDED"] = "FUNDED";
})(EAccountFundingStatus = exports.EAccountFundingStatus || (exports.EAccountFundingStatus = {}));
var EAccountFunctionality;
(function (EAccountFunctionality) {
    EAccountFunctionality["SEED_PHRASE_RECOVERY"] = "SEED_PHRASE_RECOVERY";
})(EAccountFunctionality = exports.EAccountFunctionality || (exports.EAccountFunctionality = {}));
var EGetKeysPermissionType;
(function (EGetKeysPermissionType) {
    EGetKeysPermissionType["ALL"] = "ALL";
    EGetKeysPermissionType["FULL_ACCESS"] = "FULL_ACCESS";
    EGetKeysPermissionType["FUNCTION_CALL"] = "FUNCTION_CALL";
})(EGetKeysPermissionType = exports.EGetKeysPermissionType || (exports.EGetKeysPermissionType = {}));
var EAccountIdentifierType;
(function (EAccountIdentifierType) {
    EAccountIdentifierType["IMPLICIT"] = "IMPLICIT";
    EAccountIdentifierType["NAMED"] = "NAMED";
})(EAccountIdentifierType = exports.EAccountIdentifierType || (exports.EAccountIdentifierType = {}));
var EAccountImportType;
(function (EAccountImportType) {
    EAccountImportType["SEED_PHRASE"] = "SEED_PHRASE";
    EAccountImportType["BATCH_IMPORT_PRIVATE_KEY"] = "BATCH_IMPORT_PRIVATE_KEY";
    EAccountImportType["PRIVATE_KEY"] = "PRIVATE_KEY";
})(EAccountImportType = exports.EAccountImportType || (exports.EAccountImportType = {}));
var EProfileStatus;
(function (EProfileStatus) {
    EProfileStatus["FRESH"] = "FRESH";
    EProfileStatus["SET_UP"] = "SET_UP";
})(EProfileStatus = exports.EProfileStatus || (exports.EProfileStatus = {}));
var EWalletUserVersion;
(function (EWalletUserVersion) {
    EWalletUserVersion["V202301"] = "V202301";
})(EWalletUserVersion = exports.EWalletUserVersion || (exports.EWalletUserVersion = {}));
