// @tldraw/editor/dist-cjs/lib/config/TLUserPreferences.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TLUserPreferences_exports = {};
__export(TLUserPreferences_exports, {
  USER_COLORS: () => USER_COLORS,
  defaultUserPreferences: () => defaultUserPreferences,
  getFreshUserPreferences: () => getFreshUserPreferences,
  getUserPreferences: () => getUserPreferences,
  setUserPreferences: () => setUserPreferences,
  userPrefersDarkUI: () => userPrefersDarkUI,
  userPrefersReducedMotion: () => userPrefersReducedMotion
});
module.exports = __toCommonJS(TLUserPreferences_exports);
var import_state = require("@tldraw/state");
var import_store = require("@tldraw/store");
var import_tlschema = require("@tldraw/tlschema");
var import_validate = require("@tldraw/validate");
var import_uniqueId = require("../utils/uniqueId");
const USER_DATA_KEY = "TLDRAW_USER_DATA_v3";
const userTypeValidator = import_validate.T.object({
  id: import_validate.T.string,
  name: import_validate.T.string.nullable().optional(),
  locale: import_validate.T.string.nullable().optional(),
  color: import_validate.T.string.nullable().optional(),
  isDarkMode: import_validate.T.boolean.nullable().optional(),
  animationSpeed: import_validate.T.number.nullable().optional(),
  isSnapMode: import_validate.T.boolean.nullable().optional()
});
const Versions = {
  AddAnimationSpeed: 1,
  AddIsSnapMode: 2,
  MakeFieldsNullable: 3
};
const userMigrations = (0, import_store.defineMigrations)({
  currentVersion: Versions.MakeFieldsNullable,
  migrators: {
    [Versions.AddAnimationSpeed]: {
      up: (user) => {
        return {
          ...user,
          animationSpeed: 1
        };
      },
      down: ({ animationSpeed: _, ...user }) => {
        return user;
      }
    },
    [Versions.AddIsSnapMode]: {
      up: (user) => {
        return { ...user, isSnapMode: false };
      },
      down: ({ isSnapMode: _, ...user }) => {
        return user;
      }
    },
    [Versions.MakeFieldsNullable]: {
      up: (user) => {
        return user;
      },
      down: (user) => {
        return {
          id: user.id,
          name: user.name ?? defaultUserPreferences.name,
          locale: user.locale ?? defaultUserPreferences.locale,
          color: user.color ?? defaultUserPreferences.color,
          isDarkMode: user.isDarkMode ?? defaultUserPreferences.isDarkMode,
          animationSpeed: user.animationSpeed ?? defaultUserPreferences.animationSpeed,
          isSnapMode: user.isSnapMode ?? defaultUserPreferences.isSnapMode
        };
      }
    }
  }
});
const USER_COLORS = [
  "#FF802B",
  "#EC5E41",
  "#F2555A",
  "#F04F88",
  "#E34BA9",
  "#BD54C6",
  "#9D5BD2",
  "#7B66DC",
  "#02B1CC",
  "#11B3A3",
  "#39B178",
  "#55B467"
];
function getRandomColor() {
  return USER_COLORS[Math.floor(Math.random() * USER_COLORS.length)];
}
function userPrefersDarkUI() {
  if (typeof window === "undefined") {
    return false;
  }
  return window.matchMedia?.("(prefers-color-scheme: dark)")?.matches ?? false;
}
function userPrefersReducedMotion() {
  if (typeof window === "undefined") {
    return false;
  }
  return window.matchMedia?.("(prefers-reduced-motion: reduce)")?.matches ?? false;
}
const defaultUserPreferences = Object.freeze({
  name: "New User",
  locale: (0, import_tlschema.getDefaultTranslationLocale)(),
  color: getRandomColor(),
  isDarkMode: false,
  animationSpeed: userPrefersReducedMotion() ? 0 : 1,
  isSnapMode: false
});
function getFreshUserPreferences() {
  return {
    id: (0, import_uniqueId.uniqueId)()
  };
}
function migrateUserPreferences(userData) {
  if (userData === null || typeof userData !== "object") {
    return getFreshUserPreferences();
  }
  if (!("version" in userData) || !("user" in userData) || typeof userData.version !== "number") {
    return getFreshUserPreferences();
  }
  const migrationResult = (0, import_store.migrate)({
    value: userData.user,
    fromVersion: userData.version,
    toVersion: userMigrations.currentVersion ?? 0,
    migrations: userMigrations
  });
  if (migrationResult.type === "error") {
    return getFreshUserPreferences();
  }
  try {
    userTypeValidator.validate(migrationResult.value);
  } catch (e) {
    return getFreshUserPreferences();
  }
  return migrationResult.value;
}
function loadUserPreferences() {
  const userData = typeof window === "undefined" ? null : JSON.parse(window?.localStorage?.getItem(USER_DATA_KEY) || "null") ?? null;
  return migrateUserPreferences(userData);
}
const globalUserPreferences = (0, import_state.atom)("globalUserData", loadUserPreferences());
function storeUserPreferences() {
  if (typeof window !== "undefined" && window.localStorage) {
    window.localStorage.setItem(
      USER_DATA_KEY,
      JSON.stringify({
        version: userMigrations.currentVersion,
        user: globalUserPreferences.get()
      })
    );
  }
}
function setUserPreferences(user) {
  userTypeValidator.validate(user);
  globalUserPreferences.set(user);
  storeUserPreferences();
  broadcastUserPreferencesChange();
}
const isTest = typeof process !== "undefined" && process.env.NODE_ENV === "test";
const channel = typeof BroadcastChannel !== "undefined" && !isTest ? new BroadcastChannel("tldraw-user-sync") : null;
channel?.addEventListener("message", (e) => {
  const data = e.data;
  if (data?.type === broadcastEventKey && data?.origin !== broadcastOrigin) {
    globalUserPreferences.set(migrateUserPreferences(data.data));
  }
});
const broadcastOrigin = (0, import_uniqueId.uniqueId)();
const broadcastEventKey = "tldraw-user-preferences-change";
function broadcastUserPreferencesChange() {
  channel?.postMessage({
    type: broadcastEventKey,
    origin: broadcastOrigin,
    data: {
      user: globalUserPreferences.get(),
      version: userMigrations.currentVersion
    }
  });
}
function getUserPreferences() {
  return globalUserPreferences.get();
}
